src/hotspot/share/opto/loopUnswitch.cpp

```diff
@@ -102,114 +116,204 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
+    assert(_phase->is_member(loop, unswitch_candidate), "must be inside original loop");
+    return unswitch_candidate;
   }
 
-  // Find first invariant test that doesn't exit the loop
-  IfNode* unswitch_iff = find_unswitching_candidate((const IdealLoopTree *)loop);
-  assert(unswitch_iff != nullptr, "should be at least one");
+  IfNode* create_selector_if() const {
+    const uint dom_depth = _phase->dom_depth(_original_loop_entry);
+    _phase->igvn().rehash_node_delayed(_original_loop_entry);
+    BoolNode* unswitch_candidate_bool = _unswitch_candidate->in(1)->as_Bool();
+    IfNode* selector_if = IfNode::make_with_same_profile(_unswitch_candidate, _original_loop_entry,
+                                                         unswitch_candidate_bool);
+    _phase->register_node(selector_if, _outer_loop, _original_loop_entry, dom_depth);
+    return selector_if;
+  }
 
-#ifndef PRODUCT
-  if (TraceLoopOpts) {
-    tty->print("Unswitch   %d ", head->unswitch_count()+1);
-    loop->dump_head();
+  IfProjNode* create_proj_to_loop(const PathToLoop path_to_loop) {
+    const uint dom_depth = _phase->dom_depth(_original_loop_entry);
+    IfProjNode* proj_to_loop;
+    if (path_to_loop == TRUE_PATH) {
+      proj_to_loop = new IfTrueNode(_selector);
+    } else {
+      proj_to_loop = new IfFalseNode(_selector);
+    }
+    _phase->register_node(proj_to_loop, _outer_loop, _selector, dom_depth);
+    return proj_to_loop;
   }
-#endif
 
-  C->print_method(PHASE_BEFORE_LOOP_UNSWITCHING, 4, head);
+ public:
+  IfNode* unswitch_candidate() const {
+    return _unswitch_candidate;
+  }
 
-  // Need to revert back to normal loop
-  if (head->is_CountedLoop() && !head->as_CountedLoop()->is_normal_loop()) {
-    head->as_CountedLoop()->set_normal_loop();
+  IfNode* selector() const {
+    return _selector;
   }
 
-  IfNode* invar_iff = create_slow_version_of_loop(loop, old_new, unswitch_iff, CloneIncludesStripMined);
-  ProjNode* proj_true = invar_iff->proj_out(1);
-  verify_fast_loop(head, proj_true);
+  IfTrueNode* true_path_loop_proj() const {
+    return _true_path_loop_proj;
+  }
 
-  // Increment unswitch count
-  LoopNode* head_clone = old_new[head->_idx]->as_Loop();
-  int nct = head->unswitch_count() + 1;
-  head->set_unswitch_count(nct);
-  head_clone->set_unswitch_count(nct);
+  IfFalseNode* false_path_loop_proj() const {
+    return _false_path_loop_proj;
+  }
+};
+
+// Class to unswitch the original loop and create Predicates at the new unswitched loop versions. The newly cloned loop
+// becomes the false-path-loop while original loop becomes the true-path-loop.
+class OriginalLoop : public StackObj {
+  LoopNode* const _strip_mined_loop_head;
+  IdealLoopTree* const _loop;
+  Node_List& _old_new;
+  PhaseIdealLoop* const _phase;
+
+ public:
+  OriginalLoop(IdealLoopTree* loop, Node_List& old_new)
+      : _strip_mined_loop_head(loop->_head->as_Loop()->skip_strip_mined()),
```

Comment:
Hmm. this could now be a strip mined loop node, or a regular loop node (if loop not strip mined), right?
Maybe you could just call it `_loop_head`, and leave a comment, see suggestion above.